torch/nn/parallel/distributed.py

```diff
@@ -377,8 +380,31 @@ def _check_default_group(self):
                                "init_process_group and have not passed "
                                "process_group argument to DDP constructor")
 
+    @contextmanager
+    def no_sync(self):
+        r"""
+        A context manager to disable gradient synchronizations across DDP
+        processes. Within this context, gradients will be accumulated on module
+        variables, which will later be synchronized in the first
+        forward-backward pass exiting the context.
+
+        Example::
+
+            >>> ddp = torch.nn.DistributedDataParallel(model, pg)
+            >>> with ddp.no_sync():
+            ...   for input in inputs:
+            ...     ddp(input).backward()  # no synchronization, accumulate grads
+            ... ddp(another_input).backward()  # synchronize grads
+        """
+        old_require_backward_grad_sync = self.require_backward_grad_sync
+        self.require_backward_grad_sync = False
+        yield
```

Comment:
This should be in a `try ... finally` block, because in case of an exception you will fail to restore the flag!